libraries/chain/webassembly/runtimes/eos-vm.cpp

```diff
@@ -106,9 +126,27 @@ void validate( const controller& control, const bytes& code, const wasm_config&
       EOS_ASSERT(apply_idx < std::numeric_limits<uint32_t>::max(), wasm_serialization_error, "apply not exported");
       const vm::func_type& apply_type = bkend.get_module().get_function_type(apply_idx);
       EOS_ASSERT((apply_type == vm::host_function{{vm::i64, vm::i64, vm::i64}, {}}), wasm_serialization_error, "apply has wrong type");
+
+      sync_call_supported = module_has_valid_sync_call(bkend.get_module());
+   } catch(vm::exception& e) {
+      EOS_THROW(wasm_serialization_error, e.detail());
+   }
+
+   return validate_result {
+      .sync_call_supported = sync_call_supported
+   };
+}
+
+bool is_sync_call_supported(const char* code_bytes, size_t code_size) {
+   wasm_code_ptr code_ptr((uint8_t*)code_bytes, code_size);
+   bool supported = false;
+   try {
+      eos_vm_null_backend_t<setcode_options> bkend(code_ptr, code_size, nullptr);
```

Comment:
afaict this is good now, but it's something to be mindful of in the future -- if we ever relax WASM validation via some protocol feature in the future, this could cause problems.

Reply:
What problems might be? Do you mean previously non-supported becomes supported?

Reply:
Thanks for the detailed explanation. If that happens, we will have to save `sync_call_supported` in the snapshot, so that the caller of `is_sync_call_supported()` can pass in `controller` to have access to protocol feature set. I will add a comment.